### ISSUE TITLE
"Date" scalar serialization on subscription results in conversion to string, fails internal type validation test

### ISSUE THREAD

**Issue by SirLamer** [nuwave/lighthouse#1606]
**Describe the bug**

For queries and mutations, Lighthouse converts the Date scalar to the Carbon library for convenient back-end development. For a subscription, Dates are cast to string before being passed to the Subscription handler (ex. "authorize"method). "Date" scalars are converted to UTC format like so: 2020-11-02T00:00:00.000000Z. Subscriptions must store the attribute values in memory in order to be available for the "filter" and "resolve" methods.

When later recalled for the "filter" or "resolve" methods, they then fail the type validation test within webonyx\graphql-php which is expecting a Date. The long-form

**Expected behavior/Solution**

Type coercion to Carbon instance for use within the Subscription class, and proper storage in memory so that the validation test doesn't later fail.

**Steps to reproduce**

1. Make a subscription with a "Date" argument.
2. Join subscription - this can work if Date argument is handled as string type
3. Broadcast on channel fails due to type validation error.

**Lighthouse Version**

4.17.0

**Comment by SirLamer** [comment]
Just reporting that I am finding the same thing happens with my custom "JSON" scalar - it is saved in the job queue in its parsed state (as an array) then lighthouse tries to put it through the decoder again which fails because it's expecting a JSON string.

**Comment by SirLamer** [comment]
See the issue I cross-referenced to here for some additional discussion.

TLDR: The variables end up looped through "executeQuery" twice, once on the first submission, and again when the Subscriber instance is recalled to prepare for the broadcast. So scalars outside of PHP standard types are getting parsed twice.

I am looking for a clean way to address this but I am really loss. Ideas include:
- Find a way to store the raw serialized values straight from the query in the initial Subscriber object. This way the values will be encoded to memory correctly and then when recalled will be parsed and presented to the PHP context like normal and as expected. This solution is best because it avoids side-effects from converting values back and forth (say, with a Carbon date for example).
- Convert the parsed variables back to the original serialized format as they came in the string for storage in the Subscription. This could have side-effect issues but presumably the same technique used to prepare data from Eloquent for a query or mutation response could be applied to this task.
- Somehow prime the GraphQL executeQuery to not try to parse values again. I personally don't like this one as there will be weird side-effects with data formats which don't serialize cleanly, like a Carbon date object made from just an ISO date string.

**Comment by spawnia** [comment]
@SirLamer good analysis. I think we will have to dig a bit deeper into the inner workings of `graphql-php` to either grab the unparsed values - or turn off parsing on the second run.

**Comment by SirLamer** [comment]
Remember that if you turn off parsing on the second run then some items may fail to substantiate for the broadcast event because the subscriber data is recalled from memory storage, ex. from redis.

**Comment by SirLamer** [comment]
I have tried my best to find a clean solution but I think it's gonna take a more systematic solution than what I can affect without changing lots of other code. Essentially the raw variables string needs to be passed along to the new Subscriber instantiation here: https://github.com/nuwave/lighthouse/blob/8b3065912800560a4e6f1100f72e7686b312d907/src/Subscriptions/SubscriptionResolverProvider.php#L75

Perhaps this can be carried in the [ResolveInfo ](https://github.com/webonyx/graphql-php/blob/c1e9c5eb7f4cf085a53f28b561b0803cdd83d8ee/src/Type/Definition/ResolveInfo.php) object separately from the parsed variables, then grabbed from there when creating the Subscriber.

In the meantime, for anyone else arriving here my hacky solution is to modify my custom Scalar to check for the type on the input to test for the "second loop" case, then respond accordingly. For the JSON scalar it works as a pass-through.

```php
  public function parseValue($value)
  {
    return is_string($value) ? $this->decodeJSON($value) : $value;
  }
```

I'll also have to extend the provided Date and DateTime scalars to affect a similar solution, as they will receive the "toString" output of a Carbon instance which is different than the intended format.

**Comment by SirLamer** [comment]
For the all of the date-related scalars, I was able to fix the issue using this method (one example provided but they're all the same):

```php
<?php

namespace App\GraphQL\Scalars;

use Carbon\Carbon;
use Nuwave\Lighthouse\Schema\Types\Scalars\DateTime as BaseDateTime;

class DateTime extends BaseDateTime
{
  public function toTypeCast(Carbon $date)
  {
    return $this->format($date);
  }

  protected function parse($value): Carbon
  {
    return parent::parse($value)->settings([
      'toStringFormat' => [$this, 'toTypeCast'], // PHP pass callable format
      'toJsonFormat' => [$this, 'toTypeCast'],
    ]);
  }
}
```

See Carbon docs here on string and JSON formatting:
https://carbon.nesbot.com/docs/#api-formatting
https://carbon.nesbot.com/docs/#api-json

In this way, when the Carbon object is serialized, it automatically outputs the original value which was used to make it, solving the issue for all of the date scalars.

Would you be willing to accept a PR for this? Technically it'd be a breaking change. Beware the status quo is that Laravel builds with Carbon v1 vs v2 will serialize differently but this fix also overrules both cases.

I still technically think the larger original issue in this thread remains as a separate issue to fix, as other custom scalars (like my JSON case) still don't work properly.

**Comment by GregPeden** [comment]
Update... my solution above seems to not be working anymore. I have not been able to pin down why but it seems that the subscriptions do not use a custom scalar solution anymore while resolving a subscription. The subscription connection works but when a subscription event is resolved I get the same error whether or not I am using Lighthouse with my custom Date / DateTime scalar solutions:

```js
{
    "channel": "private-lighthouse-Ns2wJimaUXU5rBHPmrqCP5bqpSAExkCm-1656541614",
    "errors": [
        {
            "message": "Variable \"$from\" got invalid value object; Expected type Date; \"Trailing data\"",
            "extensions": {
                "category": "graphql"
            },
            "trace": [
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/Schema/Types/Scalars/DateScalar.php",
                    "line": 36,
                    "call": "Nuwave\\Lighthouse\\Schema\\Types\\Scalars\\DateScalar::tryParsingDate(instance of Illuminate\\Support\\Carbon, 'GraphQL\\Error\\Error')"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/Utils/Value.php",
                    "line": 68,
                    "call": "Nuwave\\Lighthouse\\Schema\\Types\\Scalars\\DateScalar::parseValue(instance of Illuminate\\Support\\Carbon)"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/Utils/Value.php",
                    "line": 55,
                    "call": "GraphQL\\Utils\\Value::coerceValue(instance of Illuminate\\Support\\Carbon, GraphQLType: Date, instance of GraphQL\\Language\\AST\\VariableDefinitionNode, null)"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/Executor/Values.php",
                    "line": 111,
                    "call": "GraphQL\\Utils\\Value::coerceValue(instance of Illuminate\\Support\\Carbon, GraphQLType: Date, instance of GraphQL\\Language\\AST\\VariableDefinitionNode)"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
                    "line": 179,
                    "call": "GraphQL\\Executor\\Values::getVariableValues(instance of GraphQL\\Type\\Schema, instance of GraphQL\\Language\\AST\\NodeList(4), array(4))"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
                    "line": 91,
                    "call": "GraphQL\\Executor\\ReferenceExecutor::buildExecutionContext(instance of GraphQL\\Type\\Schema, instance of GraphQL\\Language\\AST\\DocumentNode, instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber, instance of Nuwave\\Lighthouse\\Schema\\Context, array(4), null, array(2), instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter)"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/Executor/Executor.php",
                    "line": 153,
                    "call": "GraphQL\\Executor\\ReferenceExecutor::create(instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter, instance of GraphQL\\Type\\Schema, instance of GraphQL\\Language\\AST\\DocumentNode, instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber, instance of Nuwave\\Lighthouse\\Schema\\Context, array(4), null, array(2))"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/GraphQL.php",
                    "line": 162,
                    "call": "GraphQL\\Executor\\Executor::promiseToExecute(instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter, instance of GraphQL\\Type\\Schema, instance of GraphQL\\Language\\AST\\DocumentNode, instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber, instance of Nuwave\\Lighthouse\\Schema\\Context, array(4), null, null)"
                },
                {
                    "file": "/var/www/html/vendor/webonyx/graphql-php/src/GraphQL.php",
                    "line": 94,
                    "call": "GraphQL\\GraphQL::promiseToExecute(instance of GraphQL\\Executor\\Promise\\Adapter\\SyncPromiseAdapter, instance of GraphQL\\Type\\Schema, instance of GraphQL\\Language\\AST\\DocumentNode, instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber, instance of Nuwave\\Lighthouse\\Schema\\Context, array(4), null, null, array(29))"
                },
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/GraphQL.php",
                    "line": 268,
                    "call": "GraphQL\\GraphQL::executeQuery(instance of GraphQL\\Type\\Schema, instance of GraphQL\\Language\\AST\\DocumentNode, instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber, instance of Nuwave\\Lighthouse\\Schema\\Context, array(4), null, null, array(29))"
                },
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/Subscriptions/SubscriptionBroadcaster.php",
                    "line": 87,
                    "call": "Nuwave\\Lighthouse\\GraphQL::executeParsedQuery(instance of GraphQL\\Language\\AST\\DocumentNode, instance of Nuwave\\Lighthouse\\Schema\\Context, array(4), instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber)"
                },
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/Subscriptions/Iterators/AuthenticatingSyncIterator.php",
                    "line": 61,
                    "call": "Nuwave\\Lighthouse\\Subscriptions\\SubscriptionBroadcaster::Nuwave\\Lighthouse\\Subscriptions\\{closure}(instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Collections/Traits/EnumeratesValues.php",
                    "line": 244,
                    "call": "Nuwave\\Lighthouse\\Subscriptions\\Iterators\\AuthenticatingSyncIterator::Nuwave\\Lighthouse\\Subscriptions\\Iterators\\{closure}(instance of Nuwave\\Lighthouse\\Subscriptions\\Subscriber, 6)"
                },
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/Subscriptions/Iterators/AuthenticatingSyncIterator.php",
                    "line": 72,
                    "call": "Illuminate\\Support\\Collection::each(instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/Subscriptions/SubscriptionBroadcaster.php",
                    "line": 97,
                    "call": "Nuwave\\Lighthouse\\Subscriptions\\Iterators\\AuthenticatingSyncIterator::process(instance of Illuminate\\Support\\Collection(7), instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/nuwave/lighthouse/src/Subscriptions/BroadcastSubscriptionJob.php",
                    "line": 46,
                    "call": "Nuwave\\Lighthouse\\Subscriptions\\SubscriptionBroadcaster::broadcast(instance of App\\GraphQL\\Subscriptions\\ActivityForChartUpdated, 'activityForChartUpdated', instance of App\\Models\\ActivityForChart)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php",
                    "line": 36,
                    "call": "Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob::handle(instance of Nuwave\\Lighthouse\\Subscriptions\\SubscriptionBroadcaster)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php",
                    "line": 40,
                    "call": "Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php",
                    "line": 93,
                    "call": "Illuminate\\Container\\Util::unwrapIfClosure(instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php",
                    "line": 37,
                    "call": "Illuminate\\Container\\BoundMethod::callBoundMethod(instance of Illuminate\\Foundation\\Application, array(2), instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php",
                    "line": 653,
                    "call": "Illuminate\\Container\\BoundMethod::call(instance of Illuminate\\Foundation\\Application, array(2), array(0), null)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php",
                    "line": 128,
                    "call": "Illuminate\\Container\\Container::call(array(2))"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
                    "line": 128,
                    "call": "Illuminate\\Bus\\Dispatcher::Illuminate\\Bus\\{closure}(instance of Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
                    "line": 103,
                    "call": "Illuminate\\Pipeline\\Pipeline::Illuminate\\Pipeline\\{closure}(instance of Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php",
                    "line": 132,
                    "call": "Illuminate\\Pipeline\\Pipeline::then(instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php",
                    "line": 120,
                    "call": "Illuminate\\Bus\\Dispatcher::dispatchNow(instance of Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob, false)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
                    "line": 128,
                    "call": "Illuminate\\Queue\\CallQueuedHandler::Illuminate\\Queue\\{closure}(instance of Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php",
                    "line": 103,
                    "call": "Illuminate\\Pipeline\\Pipeline::Illuminate\\Pipeline\\{closure}(instance of Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php",
                    "line": 122,
                    "call": "Illuminate\\Pipeline\\Pipeline::then(instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/CallQueuedHandler.php",
                    "line": 70,
                    "call": "Illuminate\\Queue\\CallQueuedHandler::dispatchThroughMiddleware(instance of Illuminate\\Queue\\Jobs\\RedisJob, instance of Nuwave\\Lighthouse\\Subscriptions\\BroadcastSubscriptionJob)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Jobs/Job.php",
                    "line": 98,
                    "call": "Illuminate\\Queue\\CallQueuedHandler::call(instance of Illuminate\\Queue\\Jobs\\RedisJob, array(2))"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php",
                    "line": 428,
                    "call": "Illuminate\\Queue\\Jobs\\Job::fire()"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php",
                    "line": 378,
                    "call": "Illuminate\\Queue\\Worker::process('redis', instance of Illuminate\\Queue\\Jobs\\RedisJob, instance of Illuminate\\Queue\\WorkerOptions)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Worker.php",
                    "line": 172,
                    "call": "Illuminate\\Queue\\Worker::runJob(instance of Illuminate\\Queue\\Jobs\\RedisJob, 'redis', instance of Illuminate\\Queue\\WorkerOptions)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php",
                    "line": 117,
                    "call": "Illuminate\\Queue\\Worker::daemon('redis', 'high,default,low', instance of Illuminate\\Queue\\WorkerOptions)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Queue/Console/WorkCommand.php",
                    "line": 101,
                    "call": "Illuminate\\Queue\\Console\\WorkCommand::runWorker('redis', 'high,default,low')"
                },
                {
                    "file": "/var/www/html/vendor/laravel/horizon/src/Console/WorkCommand.php",
                    "line": 51,
                    "call": "Illuminate\\Queue\\Console\\WorkCommand::handle()"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php",
                    "line": 36,
                    "call": "Laravel\\Horizon\\Console\\WorkCommand::handle()"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/Util.php",
                    "line": 40,
                    "call": "Illuminate\\Container\\BoundMethod::Illuminate\\Container\\{closure}()"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php",
                    "line": 93,
                    "call": "Illuminate\\Container\\Util::unwrapIfClosure(instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/BoundMethod.php",
                    "line": 37,
                    "call": "Illuminate\\Container\\BoundMethod::callBoundMethod(instance of Illuminate\\Foundation\\Application, array(2), instance of Closure)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Container/Container.php",
                    "line": 653,
                    "call": "Illuminate\\Container\\BoundMethod::call(instance of Illuminate\\Foundation\\Application, array(2), array(0), null)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php",
                    "line": 136,
                    "call": "Illuminate\\Container\\Container::call(array(2))"
                },
                {
                    "file": "/var/www/html/vendor/symfony/console/Command/Command.php",
                    "line": 298,
                    "call": "Illuminate\\Console\\Command::execute(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Illuminate\\Console\\OutputStyle)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Console/Command.php",
                    "line": 121,
                    "call": "Symfony\\Component\\Console\\Command\\Command::run(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Illuminate\\Console\\OutputStyle)"
                },
                {
                    "file": "/var/www/html/vendor/symfony/console/Application.php",
                    "line": 1015,
                    "call": "Illuminate\\Console\\Command::run(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Symfony\\Component\\Console\\Output\\ConsoleOutput)"
                },
                {
                    "file": "/var/www/html/vendor/symfony/console/Application.php",
                    "line": 299,
                    "call": "Symfony\\Component\\Console\\Application::doRunCommand(instance of Laravel\\Horizon\\Console\\WorkCommand, instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Symfony\\Component\\Console\\Output\\ConsoleOutput)"
                },
                {
                    "file": "/var/www/html/vendor/symfony/console/Application.php",
                    "line": 171,
                    "call": "Symfony\\Component\\Console\\Application::doRun(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Symfony\\Component\\Console\\Output\\ConsoleOutput)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Console/Application.php",
                    "line": 94,
                    "call": "Symfony\\Component\\Console\\Application::run(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Symfony\\Component\\Console\\Output\\ConsoleOutput)"
                },
                {
                    "file": "/var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Console/Kernel.php",
                    "line": 129,
                    "call": "Illuminate\\Console\\Application::run(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Symfony\\Component\\Console\\Output\\ConsoleOutput)"
                },
                {
                    "file": "/var/www/html/artisan",
                    "line": 38,
                    "call": "Illuminate\\Foundation\\Console\\Kernel::handle(instance of Symfony\\Component\\Console\\Input\\ArgvInput, instance of Symfony\\Component\\Console\\Output\\ConsoleOutput)"
                }
            ]
        }
    ]
}

```

**Comment by mserralta** [comment]
I was speding all afternoon looking for a mistake on my side and I see you describe perfectly the same scenario. Did you get to work with it?

Thanks!

**Comment by danbka33** [comment]
Still actual, v6.22.0
